shared/fs/row/drop-target.js

```diff
@@ -0,0 +1,81 @@
+// @flow
+import * as React from 'react'
+import {Box, Icon} from '../../common-adapters'
+import * as Styles from '../../styles'
+import {isMobile} from '../../constants/platform'
+
+type State = {|
+  showDropOverlay: boolean,
+|}
+
+type Props = {
+  children?: React.Node,
+  onAttach: Array<string> => void,
+}
+
+const DropOverlay = ({onDragLeave, onDrop}) => (
+  <Box style={styles.dropOverlayStyle} onDragLeave={onDragLeave} onDrop={onDrop}>
+    <Icon type="icon-dropping-file-48" />
+  </Box>
+)
+
+class DropTarget extends React.PureComponent<Props, State> {
+  state = {showDropOverlay: false}
+
+  _onDrop = e => {
+    if (!this._validDrag(e)) {
+      return
+    }
+    const fileList = e.dataTransfer.files
+    const paths = fileList.length ? Array.prototype.map.call(fileList, f => f.path) : []
+    this.props.onAttach(paths)
+    this.setState({showDropOverlay: false})
+  }
+
+  _validDrag = e => Array.prototype.map.call(e.dataTransfer.types, t => t).includes('Files')
+
+  _onDragOver = e => {
+    if (this._validDrag(e)) {
+      e.dataTransfer.dropEffect = 'copy'
+      this.setState({showDropOverlay: true})
+    } else {
+      e.dataTransfer.dropEffect = 'none'
+    }
+  }
+
+  _onDragLeave = e => {
+    this.setState({showDropOverlay: false})
+  }
+
+  render() {
+    const {children, onAttach} = this.props
+    return (
+      <Box style={styles.containerStyle} onDragOver={onAttach ? this._onDragOver : null}>
+        {children}
+        {this.state.showDropOverlay && <DropOverlay onDragLeave={this._onDragLeave} onDrop={this._onDrop} />}
+      </Box>
```

Comment:
If you decide on using the `children`, I think this could also be like this:

```
<>
  {children}
  {this.state.showDropOverlay && <DropOverlay onDragLeave={this._onDragLeave} onDrop={this._onDrop} />}
</>
```

The reason why I'm suggesting this is this way `Rows` would always be inside the same parent Box, and it seems it's less likely CSS goes wrong in different scenarios.